File: hub_go/internal/sync/todos.go
package sync

import "strings"

type TodoItem struct {
	Content  string
	Priority string
	Status   string
	ID       string
}

func ExtractTodoWriteTodosFromMessageContent(messageContent any) []map[string]any {
	record, ok := unwrapRoleWrappedRecordEnvelope(messageContent)
	if !ok {
		return nil
	}
	role, _ := record["role"].(string)
	if role != "agent" && role != "assistant" {
		return nil
	}
	content, ok := record["content"].(map[string]any)
	if !ok {
		return nil
	}
	contentType, _ := content["type"].(string)
	if contentType == "output" {
		if todos := extractTodosFromClaudeOutput(content); todos != nil {
			return todos
		}
	}
	if contentType == "codex" {
		if todos := extractTodosFromCodexMessage(content); todos != nil {
			return todos
		}
		if todos := extractTodosFromAcpMessage(content); todos != nil {
			return todos
		}
	}
	return nil
}

func unwrapRoleWrappedRecordEnvelope(value any) (map[string]any, bool) {
	if obj, ok := value.(map[string]any); ok {
		if isRoleWrappedRecord(obj) {
			return obj, true
		}
		if direct, ok := obj["message"].(map[string]any); ok && isRoleWrappedRecord(direct) {
			return direct, true
		}
		if data, ok := obj["data"].(map[string]any); ok {
			if msg, ok := data["message"].(map[string]any); ok && isRoleWrappedRecord(msg) {
				return msg, true
			}
		}
		if payload, ok := obj["payload"].(map[string]any); ok {
			if msg, ok := payload["message"].(map[string]any); ok && isRoleWrappedRecord(msg) {
				return msg, true
			}
		}
	}
	return nil, false
}

func isRoleWrappedRecord(obj map[string]any) bool {
	role, ok := obj["role"].(string)
	if !ok || role == "" {
		return false
	}
	_, ok = obj["content"]
	return ok
}

func extractTodosFromClaudeOutput(content map[string]any) []map[string]any {
	if content["type"] != "output" {
		return nil
	}
	data, ok := content["data"].(map[string]any)
	if !ok || data["type"] != "assistant" {
		return nil
	}
	message, ok := data["message"].(map[string]any)
	if !ok {
		return nil
	}
	blocks, ok := message["content"].([]any)
	if !ok {
		return nil
	}
	for _, block := range blocks {
		entry, ok := block.(map[string]any)
		if !ok {
			continue
		}
		if entry["type"] != "tool_use" {
			continue
		}
		name, _ := entry["name"].(string)
		if name != "TodoWrite" {
			continue
		}
		input, ok := entry["input"].(map[string]any)
		if !ok {
			continue
		}
		if todos := validateTodos(input["todos"]); todos != nil {
			return todos
		}
	}
	return nil
}

func extractTodosFromCodexMessage(content map[string]any) []map[string]any {
	if content["type"] != "codex" {
		return nil
	}
	data, ok := content["data"].(map[string]any)
	if !ok || data["type"] != "tool-call" {
		return nil
	}
	name, _ := data["name"].(string)
	if name != "TodoWrite" {
		return nil
	}
	input, ok := data["input"].(map[string]any)
	if !ok {
		return nil
	}
	return validateTodos(input["todos"])
}

func extractTodosFromAcpMessage(content map[string]any) []map[string]any {
	if content["type"] != "codex" {
		return nil
	}
	data, ok := content["data"].(map[string]any)
	if !ok || data["type"] != "plan" {
		return nil
	}
	entries, ok := data["entries"].([]any)
	if !ok {
		return nil
	}

	todos := make([]map[string]any, 0, len(entries))
	for index, entry := range entries {
		item, ok := entry.(map[string]any)
		if !ok {
			continue
		}
		contentValue, _ := item["content"].(string)
		priorityValue, _ := item["priority"].(string)
		statusValue, _ := item["status"].(string)
		if contentValue == "" || !validPriority(priorityValue) || !validStatus(statusValue) {
			continue
		}
		idValue, _ := item["id"].(string)
		if idValue == "" {
			idValue = "plan-" + itoa(index+1)
		}
		todos = append(todos, map[string]any{
			"content":  contentValue,
			"priority": priorityValue,
			"status":   statusValue,
			"id":       idValue,
		})
	}
	if len(todos) == 0 {
		return nil
	}
	return todos
}

func validateTodos(candidate any) []map[string]any {
	items, ok := candidate.([]any)
	if !ok || len(items) == 0 {
		return nil
	}
	todos := make([]map[string]any, 0, len(items))
	for _, raw := range items {
		item, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		contentValue, _ := item["content"].(string)
		priorityValue, _ := item["priority"].(string)
		statusValue, _ := item["status"].(string)
		idValue, _ := item["id"].(string)
		if contentValue == "" || !validPriority(priorityValue) || !validStatus(statusValue) {
			continue
		}
		if idValue == "" {
			idValue = strings.ReplaceAll(contentValue, " ", "-")
		}
		todos = append(todos, map[string]any{
			"content":  contentValue,
			"priority": priorityValue,
			"status":   statusValue,
			"id":       idValue,
		})
	}
	if len(todos) == 0 {
		return nil
	}
	return todos
}

func validPriority(value string) bool {
	return value == "high" || value == "medium" || value == "low"
}

func validStatus(value string) bool {
	return value == "pending" || value == "in_progress" || value == "completed"
}

func itoa(value int) string {
	if value == 0 {
		return "0"
	}
	buf := [20]byte{}
	i := len(buf)
	n := value
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

File: hub/src/sync/todos.ts
import { isObject } from '@hapi/protocol'
import { unwrapRoleWrappedRecordEnvelope } from '@hapi/protocol/messages'
import { TodoItemSchema, TodosSchema } from '@hapi/protocol/schemas'
import type { TodoItem } from '@hapi/protocol/types'

export { TodoItemSchema, TodosSchema }
export type { TodoItem }

function normalizeTodoStatus(value: unknown): TodoItem['status'] {
    if (typeof value !== 'string') return 'pending'
    const status = value.trim().toLowerCase().replace(/[\s-]+/g, '_')
    if (status === 'completed' || status === 'complete' || status === 'done') return 'completed'
    if (status === 'in_progress' || status === 'inprogress' || status === 'active' || status === 'running') {
        return 'in_progress'
    }
    return 'pending'
}

function normalizeTodoPriority(value: unknown): TodoItem['priority'] {
    if (value === 'high' || value === 'low') return value
    return 'medium'
}

function extractText(value: Record<string, unknown>): string | null {
    for (const key of ['content', 'step', 'text', 'title', 'description']) {
        const candidate = value[key]
        if (typeof candidate !== 'string') continue
        const text = candidate.trim()
        if (text) return text
    }
    return null
}

function normalizePlanEntries(candidate: unknown): TodoItem[] | null {
    if (!Array.isArray(candidate)) return null

    const todos: TodoItem[] = []
    candidate.forEach((entry, index) => {
        if (typeof entry === 'string') {
            const content = entry.trim()
            if (!content) return
            todos.push({
                content,
                priority: 'medium',
                status: 'pending',
                id: `plan-${index + 1}`
            })
            return
        }

        if (!isObject(entry)) return
        const content = extractText(entry)
        if (!content) return
        const id = typeof entry.id === 'string' && entry.id.trim()
            ? entry.id.trim()
            : `plan-${index + 1}`
        todos.push({
            content,
            priority: normalizeTodoPriority(entry.priority),
            status: normalizeTodoStatus(entry.status ?? entry.state),
            id
        })
    })

    const parsed = TodosSchema.safeParse(todos)
    return parsed.success ? parsed.data : null
}

function extractTodosFromPlanText(value: unknown): TodoItem[] | null {
    if (typeof value !== 'string') return null

    const listLines = value.split(/\r?\n/).flatMap((line) => {
        const match = line.match(/^(\s*)(?:[-*+]\s+|\d+[.)]\s+)(.+?)\s*$/)
        if (!match?.[2]) return []
        return [{ indent: match[1]?.length ?? 0, content: match[2] }]
    })

    if (listLines.length === 0) return null
    const topLevelIndent = Math.min(...listLines.map((line) => line.indent))
    return normalizePlanEntries(
        listLines
            .filter((line) => line.indent === topLevelIndent)
            .map((line) => line.content)
    )
}

function isPlanExitToolName(value: unknown): boolean {
    return value === 'ExitPlanMode' || value === 'exit_plan_mode'
}

function extractTodosFromClaudeOutput(content: Record<string, unknown>): TodoItem[] | null {
    if (content.type !== 'output') return null

    const data = isObject(content.data) ? content.data : null
    if (!data || data.type !== 'assistant') return null

    const message = isObject(data.message) ? data.message : null
    if (!message) return null

    const modelContent = message.content
    if (!Array.isArray(modelContent)) return null

    let planExitSeen = false
    let planText: string | null = null
    for (const block of modelContent) {
        if (!isObject(block)) continue
        if (block.type === 'text' && typeof block.text === 'string') {
            planText ??= block.text
            continue
        }
        if (block.type !== 'tool_use') continue
        const name = typeof block.name === 'string' ? block.name : null
        const input = 'input' in block ? (block as Record<string, unknown>).input : null
        if (name === 'TodoWrite') {
            if (!isObject(input)) continue

            const todosCandidate = input.todos
            const parsed = TodosSchema.safeParse(todosCandidate)
            if (parsed.success) {
                return parsed.data
            }
            continue
        }
        if (isPlanExitToolName(name)) {
            planExitSeen = true
            if (isObject(input) && typeof input.plan === 'string') {
                planText = input.plan
            }
        }
    }

    return planExitSeen ? extractTodosFromPlanText(planText) : null
}

function extractTodosFromCodexMessage(content: Record<string, unknown>): TodoItem[] | null {
    if (content.type !== 'codex') return null

    const data = isObject(content.data) ? content.data : null
    if (!data) return null

    if (data.type === 'plan_update') {
        return normalizePlanEntries(data.plan ?? data.update ?? data.items ?? data.steps)
    }

    if (data.type !== 'tool-call') return null

    const name = typeof data.name === 'string' ? data.name : null
    const input = 'input' in data ? (data as Record<string, unknown>).input : null
    if (name === 'update_plan') {
        return normalizePlanEntries(isObject(input) ? input.plan : null)
    }
    if (name === 'TodoWrite') {
        if (!isObject(input)) return null

        const todosCandidate = input.todos
        const parsed = TodosSchema.safeParse(todosCandidate)
        return parsed.success ? parsed.data : null
    }
    if (isPlanExitToolName(name)) {
        return extractTodosFromPlanText(isObject(input) ? input.plan : null)
    }

    return null
}

function extractTodosFromAcpMessage(content: Record<string, unknown>): TodoItem[] | null {
    if (content.type !== 'codex') return null

    const data = isObject(content.data) ? content.data : null
    if (!data || data.type !== 'plan') return null

    return normalizePlanEntries(data.entries)
}

export function extractSessionTodosFromMessageContent(messageContent: unknown): TodoItem[] | null {
    const record = unwrapRoleWrappedRecordEnvelope(messageContent)
    if (!record) return null

    if (record.role !== 'agent' && record.role !== 'assistant') return null

    if (!isObject(record.content) || typeof record.content.type !== 'string') return null

    return extractTodosFromClaudeOutput(record.content)
        ?? extractTodosFromCodexMessage(record.content)
        ?? extractTodosFromAcpMessage(record.content)
}
